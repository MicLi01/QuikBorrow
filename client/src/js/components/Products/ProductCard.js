import React from "react";
import styled from 'styled-components';
import { useSelector } from 'react-redux'
import { Card, CardMedia, CardContent, Typography, Button, Tooltip, ButtonGroup } from '@material-ui/core'
import { useHistory } from 'react-router-dom';
import green from '@material-ui/core/colors/green'
import grey from '@material-ui/core/colors/grey'

const SpaceBetween = styled.div`
  display: flex;
  justify-content: space-between;
`

// http://hackingui.com/front-end/a-pure-css-solution-for-multiline-text-truncation/
const TruncatedText = styled.p`
  overflow: hidden;
  position: relative; 
  line-height: 1.2em;
  max-height: 3.6em; 
  text-align: justify;  
  padding-right: -1em;
  padding-right: 1em;
  &::before {
    content: '...';
    position: absolute;
    right: 0;
    bottom: 0;
  }
  &::after {
    content: '';
    position: absolute;
    right: 0;
    width: 1em;
    height: 1em;
    margin-top: 0.2em;
    background: white;
  }
`

const FlexCard = styled(Card)`
  display: flex;
  height: 200px;
`

const FlexCardImage = styled(CardMedia)`
  width: 200px;
  height: auto;
  flex-shrink: 0;
`

const FlexCardContent = styled(CardContent)`
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`

const ProductCard = props => {
  const { id, name, user, description, availability, imagePath, borrower } = props;
  const { user: loggedInUser } = useSelector(state => state.auth)
  const history = useHistory()
  
  const availabilityText = user === loggedInUser.id
    ? <span style={{ color: green[500], fontWeight: 600 }}>You own this item</span>
    : !availability
      ? <span style={{ color: grey[500] }}>Currently being lended out</span>
      : borrower === loggedInUser.id
        ? <span style={{ color: green[500] }}>You're currently borrowing this item</span>
        : <span>Available to borrow</span>

  return (
    <FlexCard>
      <FlexCardImage image={imagePath || 'https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/No_picture_available.png/602px-No_picture_available.png'} title='image' />
      <FlexCardContent>
        <div>
          <SpaceBetween>
            <Typography component='h6' variant='h6'>
              {name}
            </Typography>
            <Typography variant='body2'>
              {availabilityText}
            </Typography>
          </SpaceBetween>
          <TruncatedText>
            {description}
          </TruncatedText>
        </div>
        <SpaceBetween>
          {availability
            ? <div />
            : <Button variant='outlined'>
              Get Notified
            </Button>
          }
          <Button variant='outlined' onClick={e => history.push(`/products/${id}`)}>
              Details
            </Button>
        </SpaceBetween>
      </FlexCardContent>
      
    </FlexCard>
  );
};

export default ProductCard;
