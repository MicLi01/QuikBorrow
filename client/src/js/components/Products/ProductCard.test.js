import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard from './ProductCard';

describe('ProductCard', () => {
	it('should render a ProductCard', () => {
		const component = renderer.create(<ProductCard />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});