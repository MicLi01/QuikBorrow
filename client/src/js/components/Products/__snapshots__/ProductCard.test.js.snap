// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should render a ProductCard 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiCard-root sc-htpNat eDgLaq MuiPaper-rounded"
>
  <div
    className="MuiCardMedia-root sc-bxivhb gVKxpq"
    style={
      Object {
        "backgroundImage": "url(\\"https://upload.wikimedia.org/wikipedia/commons/thumb/f/fc/No_picture_available.png/602px-No_picture_available.png\\")",
      }
    }
    title="image"
  />
  <div
    className="MuiCardContent-root sc-ifAKCX hCIaTb"
  >
    <div>
      <div
        className="sc-bdVaJa hPftyW"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6"
        />
        <p
          className="MuiTypography-root MuiTypography-body2"
        >
          <span
            style={
              Object {
                "color": "#4caf50",
                "fontWeight": 600,
              }
            }
          >
            You own this item
          </span>
        </p>
      </div>
      <p
        className="sc-bwzfXH qHaOC"
      />
    </div>
    <div
      className="sc-bdVaJa hPftyW"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Get Notified
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Details
        </span>
      </button>
    </div>
  </div>
</div>
`;
